### DISCUSSION TITLE
How should xdist be gracefully exited mid run when some exception occurs

### DISCUSSION THREAD

**Discussion by symonk** [pytest-dev/pytest#7869] [Q&A]
I have a scenario where I am applying some logic during an xdist -n=X run on the master node and sometimes explicitly on GW0 (tho i might refactor that to a plugin and have master handle it).  In the event that something has gone wrong, how should I *gracefully* exit pytest/xdist across all workers?

Is there a simple means of doing this?  I'm achieving it through some hacky methods at the moment and I'm wondering if there is actually a proper mechanism for handling it

**Comment by nicoddemus** [comment]
I didn't try it, but maybe `pytest.exit()` will do the trick? Don't remember the details right now.

**Reply by symonk** [reply]
thanks, will have a try with it

**Comment by Raphasha27** [comment]
Use `pytest-xdist` with `--max-worker-restart` and a timeout plugin:

```bash
pytest -n auto --max-worker-restart=3 --timeout=60
```

For graceful cleanup on exceptions, wrap your test logic and call `pytest.exit()`:

```python
# conftest.py
import pytest
import sys

def pytest_sessionfinish(session, exitstatus):
    if exitstatus != 0:
        print("Gracefully shutting down xdist workers...")
        # workers drain naturally after sessionfinish
```

If a single worker crashes, xdist reassigns its tests to other workers. To handle long-running hangs, combine with `pytest-timeout`:

```python
# pytest.ini
[pytest]
timeout = 120
timeout_method = signal  # or thread
```

This kills hung workers and lets remaining workers continue. For full control, use `--dist loadscope` to group tests by class/module so a single failure doesn't fragment test groups.
